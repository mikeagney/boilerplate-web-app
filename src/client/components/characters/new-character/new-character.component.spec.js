import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import NewCharacter from './new-character.component';

describe('NewCharacter component', () => {
  const defaultProps = {
    characterIds: ['a1a', 'b2b'],
    setSelected: jest.fn(),
    addCharacter: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('will render the expected new character form', () => {
    // Arrange
    const props = defaultProps;

    // Act
    const wrapper = shallow(<NewCharacter {...props} />);

    // Assert
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('will update the edit field when name edit is committed', () => {
    // Arrange
    const props = defaultProps;
    const wrapper = shallow(<NewCharacter {...props} />);

    // Act
    wrapper
      .find('.name-clickable-edit')
      .props()
      .setText('Changed Character');

    // Assert
    expect(wrapper.find('.name-clickable-edit').props().text).toEqual('Changed Character');
  });

  it('will reset to default name and reset selected character ID on cancel', () => {
    // Arrange
    const props = defaultProps;
    const wrapper = shallow(<NewCharacter {...props} />);
    wrapper
      .find('.name-clickable-edit')
      .props()
      .setText('Changed Character');

    // Act
    wrapper.find('.cancel-button').simulate('click');

    // Assert
    expect(wrapper.find('.name-clickable-edit').props().text).toEqual('New Character');
    expect(props.setSelected).toHaveBeenCalledTimes(1);
    expect(props.setSelected).toHaveBeenCalledWith('a1a');
  });

  it('will set selected character ID to empty on cancel when there are no character IDs', () => {
    // Arrange
    const props = { ...defaultProps, characterIds: [] };
    const wrapper = shallow(<NewCharacter {...props} />);

    // Act
    wrapper.find('.cancel-button').simulate('click');

    // Assert
    expect(props.setSelected).toHaveBeenCalledTimes(1);
    expect(props.setSelected).toHaveBeenCalledWith('');
  });

  it('will reset to default name and invoke addCharacter on save', () => {
    // Arrange
    const props = defaultProps;
    const wrapper = shallow(<NewCharacter {...props} />);
    wrapper
      .find('.name-clickable-edit')
      .props()
      .setText('Changed Character');

    // Act
    wrapper.find('.save-button').simulate('click');

    // Assert
    expect(wrapper.find('.name-clickable-edit').props().text).toEqual('New Character');
    expect(props.addCharacter).toHaveBeenCalledTimes(1);
    expect(props.addCharacter).toHaveBeenCalledWith(expect.any(String), {
      name: 'Changed Character',
    });
  });
});
