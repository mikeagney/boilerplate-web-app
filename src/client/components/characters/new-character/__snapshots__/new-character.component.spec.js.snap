// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewCharacter component will render a spinner in place of the edit buttons when creating 1`] = `
<Card
  as="div"
  body={false}
  className="character"
>
  <CardHeader
    as="h5"
    className="bg-primary text-white"
  >
    <ClickableEdit
      className="name-clickable-edit"
      loading={false}
      setText={[Function]}
      text="New Character"
    />
  </CardHeader>
  <CardBody>
    <CardText>
      Creating a new character
    </CardText>
  </CardBody>
  <CardFooter>
    <Bootstrap(Spinner)
      animation="border"
      className="test-creating-spinner"
      role="status"
      variant="secondary"
    >
      <span
        className="sr-only"
      >
        Creating...
      </span>
    </Bootstrap(Spinner)>
  </CardFooter>
</Card>
`;

exports[`NewCharacter component will render the expected new character form 1`] = `
<Card
  as="div"
  body={false}
  className="character"
>
  <CardHeader
    as="h5"
    className="bg-primary text-white"
  >
    <ClickableEdit
      className="name-clickable-edit"
      loading={false}
      setText={[Function]}
      text="New Character"
    />
  </CardHeader>
  <CardBody>
    <CardText>
      Creating a new character
    </CardText>
  </CardBody>
  <CardFooter>
    <Button
      active={false}
      className="save-button mr-1"
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="primary"
    >
      Save
    </Button>
    <Button
      active={false}
      className="cancel-button"
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      Cancel
    </Button>
  </CardFooter>
</Card>
`;
